### ISSUE TITLE
Felix sometimes stops updating iptables.

### ISSUE THREAD

**Issue by Dirbaio** [projectcalico/felix#1522]
Sometimes felix stops updating iptables, breaking networking for new deployed containers.
This happens every month or so on our beta cluster. 

## Expected Behavior

Felix shouldn't stop updating the iptables.

## Current Behavior

When the bug happens, iptables stops getting updated with new policies or newly started containers.

When starting a new container, the route to it gets created:

    root@nodes-kr1m:~$ ip route | grep 100.92.223.207
    100.92.223.207 dev cali59ef8132c6c  scope link 

but if you search the interface name in iptablres, it's not there:

    root@nodes-kr1m:~# iptables-save | grep cali59ef8132c6c
    # shows no output.

Here are the calico-node container logs. Note that around `14:03` all logs about iptables and ipset stop happening, which is the time calico stops working.
https://gist.github.com/Dirbaio/7d529d7a4e60953187d1ca5d7d3484bc

There doesn't seem to be any bad error message in the logs.

## Possible Solution

As a workaround, restarting calico-node container fixes it.

## Steps to Reproduce (for bugs)

1. Deploy calico
2. Wait
3. The bug may happen after months of having everyting up.

## Your Environment

* And/Or, `calico/node` container version (if running Felix in a container): quay.io/calico/node:v1.3.0
* Which orchestrator are you using (e.g. Kubernetes, OpenStack, Docker, Mesos): Kubernetes
* Orchestrator version (if used): 1.5.2
* Etcd version (if using etcd datastore driver): gcr.io/google_containers/etcd:2.2.1
* Operating System and version: Ubuntu 16.04

**Comment by Dirbaio** [comment]
Killing felix with SIGABRT when it's in this hung state results in this goroutine dump: https://gist.github.com/Dirbaio/39d18dfb826a34edb627a5c401cbf78f

**Comment by fasaxc** [comment]
@Dirbaio thanks for filing an issue with excellent diagnostics.  The goroutine dump should be just what we need to unpick this issue.

**Comment by fasaxc** [comment]
This might be the root cause. Looks like the netlink library has been waiting for 66 minutes to receive a response from the kernel:
```
goroutine 10672222 [syscall, 66 minutes]:
syscall.Syscall6(0x2d, 0x4, 0xc420d3e000, 0x1000, 0x0, 0xc420cb9268, 0xc420cb925c, 0x2b, 0x7f49c76b1a00, 0x450a40)
	/usr/local/go/src/syscall/asm_linux_amd64.s:44 +0x5
syscall.recvfrom(0x4, 0xc420d3e000, 0x1000, 0x1000, 0x0, 0xc420cb9268, 0xc420cb925c, 0x0, 0xc420cb9294, 0xc42043ac00)
	/usr/local/go/src/syscall/zsyscall_linux_amd64.go:1712 +0x99
syscall.Recvfrom(0x4, 0xc420d3e000, 0x1000, 0x1000, 0x0, 0x1000, 0x4, 0x200d600, 0xc420f4c3c0, 0x0)
	/usr/local/go/src/syscall/syscall_unix.go:252 +0xaf
github.com/projectcalico/felix/vendor/github.com/vishvananda/netlink/nl.(*NetlinkSocket).Receive(0xc420755260, 0xc49c9866d2, 0x0, 0x0, 0x0, 0x40eb82)
	/go/src/github.com/projectcalico/felix/vendor/github.com/vishvananda/netlink/nl/nl_linux.go:613 +0x9b
github.com/projectcalico/felix/vendor/github.com/vishvananda/netlink/nl.(*NetlinkRequest).Execute(0xc420137ae0, 0x0, 0x12f0000, 0x0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/projectcalico/felix/vendor/github.com/vishvananda/netlink/nl/nl_linux.go:410 +0x181
github.com/projectcalico/felix/vendor/github.com/vishvananda/netlink.execGetLink(0xc420137ae0, 0xc420de7160, 0x1, 0x1, 0x2)
	/go/src/github.com/projectcalico/felix/vendor/github.com/vishvananda/netlink/link_linux.go:1018 +0x3f
github.com/projectcalico/felix/vendor/github.com/vishvananda/netlink.(*Handle).LinkByName(0x208ab20, 0xc420e24540, 0xf, 0x7f49c76eb000, 0x0, 0x407835, 0xc420cb9640)
	/go/src/github.com/projectcalico/felix/vendor/github.com/vishvananda/netlink/link_linux.go:958 +0x215
github.com/projectcalico/felix/vendor/github.com/vishvananda/netlink.LinkByName(0xc420e24540, 0xf, 0x13656ad, 0xc420cb9658, 0xc420cb9650, 0xc420e9ad80)
	/go/src/github.com/projectcalico/felix/vendor/github.com/vishvananda/netlink/link_linux.go:941 +0x41
github.com/projectcalico/felix/routetable.realDataplane.LinkByName(0xc42000c1c0, 0xc420e24540, 0xf, 0xc420de7130, 0xc420755230, 0xc420137a00, 0x1)
	/go/src/github.com/projectcalico/felix/routetable/dataplane_shim.go:46 +0x35
github.com/projectcalico/felix/routetable.(*RouteTable).syncRoutesForLink(0xc4202c0620, 0xc420e24540, 0xf, 0x0, 0x0)
	/go/src/github.com/projectcalico/felix/routetable/route_table.go:258 +0x519
github.com/projectcalico/felix/routetable.(*RouteTable).Apply.func1(0x148f6c0, 0xc420e24c10, 0xc420cb9dd0, 0xc420498990)
	/go/src/github.com/projectcalico/felix/routetable/route_table.go:178 +0x112
github.com/projectcalico/felix/set.mapSet.Iter(0xc420bce540, 0xc420e24f20)
	/go/src/github.com/projectcalico/felix/set/set.go:101 +0xaf
github.com/projectcalico/felix/routetable.(*RouteTable).Apply(0xc4202c0620, 0x0, 0x0)
	/go/src/github.com/projectcalico/felix/routetable/route_table.go:200 +0xa4
github.com/projectcalico/felix/intdataplane.(*InternalDataplane).apply.func2(0xc420222f00, 0xc420930800, 0xc4202c0620)
	/go/src/github.com/projectcalico/felix/intdataplane/int_dataplane.go:740 +0x2f
created by github.com/projectcalico/felix/intdataplane.(*InternalDataplane).apply
	/go/src/github.com/projectcalico/felix/intdataplane/int_dataplane.go:746 +0x2fc
```

**Comment by fasaxc** [comment]
@Dirbaio you may be interested in our 2.4 release, which includes support for liveness/readiness checking in k8s.  With that enabled, I think k8s would have spotted that Felix had hung and restarted it in this case.

**Comment by fasaxc** [comment]
@Dirbaio 

- Can you elaborate on your setup at all?  It's be good to try to narrow down under what circumstances this is hittable.
- Are you running on VMs?
- In which cloud, if any?
- Any non-standard kernel?
- Are the machines running hot?
- How many pods per machine?
- How big are the machines?
- How often do you hit this and at what scale?

**Comment by fasaxc** [comment]
Looks like we could set a timeout by using a netlink Handle instead of relying on the default one: https://github.com/vishvananda/netlink/issues/254#issuecomment-321043032

**Comment by Dirbaio** [comment]
@fasaxc: 
I'm running on GCE VMs. Ubuntu 16.04, kernel 4.4.0-66-generic. No custom patches.
k8s 1.5.2 deployed with kops. IPIP, outgoing NAT, BGP. 

Here's the yaml we use (based off the oficial ones with very minor changes): https://gist.github.com/Dirbaio/11b177b3a46ba4d8fb0edc74f6f51b0e

I see this around once every 1-2 months in our alpha/beta clusters. I've only seen this in production once ever. (we've been running this setup since March). When it happens, it affects only a single node (not all of them at once).

I'm not sure of the reason of it being more frequent on alpha/beta, the only differences I can think of are these:

1) cluster size:
Production has 4 nodes. n1-standard-4 (4 vCPUs, 15 GB memory). 
Alpha/beta have 2 nodes each, n1-standard-2 (2 vCPUs, 7.5 GB memory).
Both have 10-15 pods per machine, not a lot.

2) deploy frquency
We deploy stuff  VERY frequently on alpha (ci deploys every commit in the repo!), somewhat frequently in beta, and ~once a week in prod. I don't  think we have this bug more frequently on alpha than beta, but it happens rarely so it may be just bad luck? 

3) load, obviously. Production has real people using it, alpha/beta sit mostly idle.

4) Prod cluster is split in 3 zones, alpha/beta clusters are just in 1 zone.

**Comment by fasaxc** [comment]
Thanks @Dirbaio, the fact that you see it more often in alpha/beta suggests churn is a factor.  Otherwise, your set-up sounds similar to a lot of our testing (where we typically use same OS on GCE on n1-standard-4s).  Odd that we don't seem to see this even on a very large scale test (1000-nodes).  If it happens once a month for you, then a 1000-node test should see it fairly quickly.

In any case, I'm working on a patch to add a timeout to that netlink request under #1526.  That should make Felix robust to whatever the underlying problem is.

**Comment by Dirbaio** [comment]
Adding a timeout sounds like a great solution :)

It happened again today in ouy beta cluster -- it looks like the cause is the same, but I got the goroutine dump anyway, just in case it's useful: https://gist.github.com/Dirbaio/60699d818c1154ed884393b0523c329a

**Comment by fasaxc** [comment]
Agreed, looks like the same stack.  Looks like my fix will miss Calico 2.5.0 (which is an unusually small/fast release) but it should be in Calico 2.6 (and we may backport it to 2.5.1, say).
